components/scream/src/share/property_checks/mass_and_energy_conservation_check.cpp

```diff
@@ -0,0 +1,336 @@
+#include "share/property_checks/mass_and_energy_conservation_check.hpp"
+#include "physics/share/physics_constants.hpp"
+#include <iomanip>
+
+namespace scream
+{
+
+MassAndEnergyConservationCheck::MassAndEnergyConservationCheck (const std::shared_ptr<const AbstractGrid>& grid,
+                                                                const std::shared_ptr<const Field>&        pseudo_density_ptr,
+                                                                const std::shared_ptr<const Field>&        ps_ptr,
+                                                                const std::shared_ptr<const Field>&        phis_ptr,
+                                                                const std::shared_ptr<const Field>&        horiz_winds_ptr,
+                                                                const std::shared_ptr<const Field>&        T_mid_ptr,
+                                                                const std::shared_ptr<const Field>&        qv_ptr,
+                                                                const std::shared_ptr<const Field>&        qc_ptr,
+                                                                const std::shared_ptr<const Field>&        qr_ptr,
+                                                                const std::shared_ptr<const Field>&        qi_ptr,
+                                                                const std::shared_ptr<const Field>&        vapor_flux_ptr,
+                                                                const std::shared_ptr<const Field>&        water_flux_ptr,
+                                                                const std::shared_ptr<const Field>&        ice_flux_ptr,
+                                                                const std::shared_ptr<const Field>&        heat_flux_ptr)
+  : m_grid (grid)
+  , m_dt (std::nan(""))
+  , m_tol (std::numeric_limits<Real>::max())
+{  
+  m_num_cols = m_grid->get_num_local_dofs();
+  m_num_levs = m_grid->get_num_vertical_levels();
+
+  m_current_mass   = view_1d<Real> ("current_total_water",  m_num_cols);
+  m_current_energy = view_1d<Real> ("current_total_energy", m_num_cols);
+
+  m_fields["pseudo_density"] = pseudo_density_ptr;
+  m_fields["ps"]             = ps_ptr;
+  m_fields["phis"]           = phis_ptr;
+  m_fields["horiz_winds"]    = horiz_winds_ptr;
+  m_fields["T_mid"]          = T_mid_ptr;
+  m_fields["qv"]             = qv_ptr;
+  m_fields["qc"]             = qc_ptr;
+  m_fields["qr"]             = qr_ptr;
+  m_fields["qi"]             = qi_ptr;
+  m_fields["vapor_flux"]     = vapor_flux_ptr;
+  m_fields["water_flux"]     = water_flux_ptr;
+  m_fields["ice_flux"]       = ice_flux_ptr;
+  m_fields["heat_flux"]      = heat_flux_ptr;
+
+  // Assert that all fields needed for mass and energy computation are not null.
+  const bool all_computation_fields_exist =
```

Comment:
Might be just me, but given that shared_ptr has the `operator bool ()` overload, having simply
```
const bool all_valid =
	pseudo_density_ptr &&  ps_ptr &&	phis_ptr && horiz_winds_ptr &&
	T_mid_ptr && ...
```
and doing away with `is_field_null` would be more readable.